### PR TITLE
Views grid

### PR DESCRIPTION
**Pull request by docwilmot**
Fixes https://github.com/backdrop/backdrop-issues/issues/479
Replaces https://github.com/backdrop/backdrop/issues/1014

### CODE REVIEW

File: core/modules/views/plugins/views_plugin_style_grid.inc [hunk 1]
```diff
@@ -12,17 +12,20 @@
...
-    $options['caption'] = array('default' => '');
-
+    $options['wrapper_classes'] = array('default' => '');
+    $options['column_classes'] = array('default' => 'views-column');
+    $options['first_column_classes'] = array('default' => 'views-column-first');
+    $options['last_column_classes'] = array('default' => 'views-column-last');
+    $options['row_classes'] = array('default' => 'views-row');
+    $options['first_row_classes'] = array('default' => 'views-row-first');
+    $options['last_row_classes'] = array('default' => 'views-row-last');
+    $options['default_classes'] = array('default' => 0);
```
Comment (klonos) [inline core/modules/views/plugins/views_plugin_style_grid.inc:29]: Excuse newbie questions but why not `array('default' => '');` here?

File: core/modules/views/plugins/views_plugin_style_grid.inc [hunk 2]
```diff
@@ -12,17 +12,20 @@
  */
 class views_plugin_style_grid extends views_plugin_style {
   /**
-   * Set default options
+   * Set default options.
    */
   function option_definition() {
     $options = parent::option_definition();
-
     $options['columns'] = array('default' => '4');
```
Comment (klonos) [inline core/modules/views/plugins/views_plugin_style_grid.inc:19]: Why not change this to `$options['columns'] = array('default' => 4);` (numeric `4` instead of string), so that we can use `'#type' => 'number',` when defining the `$form['columns']` array?

File: core/modules/views/plugins/views_plugin_style_grid.inc [hunk 3]
```diff
@@ -31,41 +34,105 @@ class views_plugin_style_grid extends views_plugin_style {
...
+    );
+    $form['wrapper_classes'] = array(
       '#type' => 'textfield',
-      '#title' => t('Short description of table'),
-      '#description' => t('Include a caption for better accessibility of your table.'),
-      '#default_value' => $this->options['caption'],
+      '#title' => t('Wrapper'),
+      '#default_value' => $options['wrapper_classes'],
+      '#fieldset' => 'grid',
+      '#prefix' => '<div class="views-left-30">',
```
Comment (klonos) [inline core/modules/views/plugins/views_plugin_style_grid.inc:79]: ...what is this `30`?? And also, if the `#prefix`/`#suffix` are the same for these form elements, then why not use a `$prefix` and a `$suffix` variable so we can change it easily in the future?

**Comment by klonos** [comment]
...PR sandbox not properly created (starts at the installation stage). @docwilmot can you please close and re-open to get a fresh one?

File: core/modules/views/templates/views-view-grid.tpl.php [hunk 4]
```diff
@@ -19,20 +19,26 @@
...
+    <div class="<?php print trim($item['classes']); ?>">
+      <?php print $item['content']; ?>
+    </div>
+  <?php endforeach; ?>
+  </div>
+<?php endforeach; ?>
+<?php else : ?>
+  <?php foreach ($rows as $row_id => $row) : ?>
+    <div class="<?php print trim($row_classes[$row_id]); ?>">
+  <?php foreach ($row as $item_id => $item) : ?>
```
Comment (klonos) [inline core/modules/views/templates/views-view-grid.tpl.php:36]: This second `foreach` is nested inside the previous one. Shouldn't it be intended 2 spaces inside the preceding `div`?

File: core/modules/views/templates/views-view-grid.tpl.php [hunk 5]
```diff
@@ -19,20 +19,26 @@
...
-</table>
+  <div class="<?php print $classes; ?>">
+<?php if ($options['alignment'] == 'vertical') : ?>
+<?php foreach ($columns as $column_id => $column) : ?>
+  <div class="testerclass <?php print trim($column_classes[$column_id]); ?>">
+  <?php foreach ($column as $item_id => $item) : ?>
+    <div class="<?php print trim($item['classes']); ?>">
+      <?php print $item['content']; ?>
+    </div>
+  <?php endforeach; ?>
```
Comment (klonos) [inline core/modules/views/templates/views-view-grid.tpl.php]: This one needs 2 space indentation.

File: core/modules/views/plugins/views_plugin_style_grid.inc (hunk 2 above)
Comment (docwilmot) [reply core/modules/views/plugins/views_plugin_style_grid.inc:19]: Yep, youre right. This was ported before we had number fields. Will need to change the validator as well.

File: core/modules/views/plugins/views_plugin_style_grid.inc (hunk 1 above)
Comment (docwilmot) [reply core/modules/views/plugins/views_plugin_style_grid.inc:29]: `$options['default_classes']` is a Boolean. See `if (!$options['default_classes'])` in `template_preprocess_views_view_grid()`

File: core/modules/views/templates/views-view-grid.tpl.php (hunk 4 above)
Comment (docwilmot) [reply core/modules/views/templates/views-view-grid.tpl.php:36]: Yes

File: core/modules/views/plugins/views_plugin_style_grid.inc (hunk 3 above)
Comment (docwilmot) [reply core/modules/views/plugins/views_plugin_style_grid.inc:79]: `views-left-30` is actually a CSS declaration in `views_ui.admin.css`: 
```
  .views-left-30 {
    float: left;
    width: 30%;
  }
```

File: core/modules/views/templates/views-view-grid.tpl.php [hunk 6]
```diff
@@ -19,20 +19,26 @@
...
-          </td>
-        <?php endforeach; ?>
-      </tr>
-    <?php endforeach; ?>
-  </tbody>
-</table>
+  <div class="<?php print $classes; ?>">
+<?php if ($options['alignment'] == 'vertical') : ?>
+<?php foreach ($columns as $column_id => $column) : ?>
+  <div class="testerclass <?php print trim($column_classes[$column_id]); ?>">
```
Comment (docwilmot) [inline core/modules/views/templates/views-view-grid.tpl.php:25]: Reminder, this must go.

File: core/modules/views/css/views.css [hunk 7]
```diff
@@ -56,3 +56,7 @@
 .view .progress-disabled {
   float: none;
 }
+
+.views-grid .views-row {
+  display: flex;
```
Comment (wesruv) [inline core/modules/views/css/views.css:61]: Needs to be run through autoprefixer, check to see if there's a plugin for your code editor of choice, or go to autoprefixer.github.io

File: core/modules/views/plugins/views_plugin_style_grid.inc (hunk 3 above)
Comment (wesruv) [reply core/modules/views/plugins/views_plugin_style_grid.inc:79]: Unfortunate holdover from terrible CSS :(

File: core/modules/views/plugins/views_plugin_style_grid.inc [hunk 8]
```diff
@@ -31,41 +35,111 @@ class views_plugin_style_grid extends views_plugin_style {
...
+      '#weight' => 80,
+      '#collapsible' => TRUE,
+      '#collapsed' => TRUE,
+    );
+    $form['column_classes'] = array(
+      '#type' => 'textfield',
+      '#title' => t('Every Column'),
+      '#default_value' => $options['column_classes'],
+      '#fieldset' => 'column_class_wrapper',
+      '#prefix' => '<div class="views-left-30">',
```
Comment (wesruv) [inline core/modules/views/plugins/views_plugin_style_grid.inc:94]: Thinking it may be worth it to write some CSS for this, the views UI CSS is just abysmal. I can help with that.

File: core/modules/views/templates/views.theme.inc [hunk 9]
```diff
@@ -676,126 +676,115 @@ function template_preprocess_views_view_table(&$variables) {
...
+        $classes = explode(' ', $options['column_classes']);
+        if (!$options['default_classes']) {
+          $classes[] = 'views-column-' . ($col_id + 1);
+        }
+        $rows[$row_id][$col_id] = array(
+          'classes' => implode(' ', $classes),
+          'content' => $variables['rows'][current($keys)],
+        );
+        if ($options['automatic_width']) {
+          $rows[$row_id][$col_id]['col_style'] = array('style' => 'flex: ' . (100 / $columns) . '%;');
```
Comment (wesruv) [inline core/modules/views/templates/views.theme.inc:714]: I would _really_ like to avoid inline styles, but I don't quite understand the scope of the problem. It's plausible that inline styles are the best solution.

Could this be updated to use the Bootstrap grid in core?

File: core/modules/views/plugins/views_plugin_style_grid.inc (hunk 8 above)
Comment (wesruv) [reply core/modules/views/plugins/views_plugin_style_grid.inc:94]: The '.fieldset-wrapper' on these fields needs a `clearfix` class, or clearfix styles should be added to it's selector in a custom stylesheet. I'm guessing there's not a sane way to put another class on `div.fieldset-wrapper`?

See without cleafix and with:
image

**Comment by wesruv** [comment]
The other issue I'm seeing (which could be solved if we use Bootstrap grid) is that there isn't any padding in between the grid elements. Could also come from custom styles in views.css (where the `display: flex` is currently)

File: core/modules/views/css/views.css (hunk 7 above)
Comment (wesruv) [reply core/modules/views/css/views.css:61]: Another thing with this, this will need a fallback for browsers that don't support flexbox. I can help with that as well.

**Comment by jenlampton** [comment]
Now that we have the bootstrap grid in core as a `#library`, it would likely be better to include that and just use the standard classes. No additional CSS would be necessary, inline or otherwise (not sure how much other stuff might have to change...)

edit: I'm currently doing a theme-override of views-view-grid in all my Backdrop sites to add the necessary classes. Would it be useful for me to post that here?

edit2: Here's one from a Drupal 7 site, still looking for a backdrop example.
```
/**
 * Prepares variables for views-view-grid.tpl.php
 */
function cfthree_preprocess_views_view_grid(&$variables) {
  $view     = $variables['view'];
  $cols     = $view->style_plugin->options['columns'];
  $rows     = $variables['rows'];

  // These views have the columns stay wider at smaller screensizes.
  $sm_grid_views = array('topics');

  if (in_array($view->name, $sm_grid_views)) {
    $column_classes = array(
      1 => 'col-sm-12',
      2 => 'col-sm-6',
      3 => 'col-sm-4',
      4 => 'col-sm-3',
      5 => 'col-sm-5ths',
      6 => 'col-sm-2',
    );
  }
  else {
    $column_classes = array(
      1 => 'col-md-12',
      2 => 'col-md-6',
      3 => 'col-md-4',
      4 => 'col-md-3',
      5 => 'col-md-5ths',
      6 => 'col-md-2',
    );
  }

  $col_class = $column_classes[$cols];

  // Apply the radix classes
  foreach ($rows as $row_number => $row) {
    $variables['row_classes'][$row_number] .= ' row';
    $variables['row_classes'][$row_number] .= ' row-fluid';
    foreach ($rows[$row_number] as $column_number => $item) {
      $variables['column_classes'][$row_number][$column_number] .= ' ' . $col_class;
    }
  }

  $variables['class'] .= ' container-fluid';
}
```
and the template file:
```
<?php if (!empty($title)) : ?>
  <h3><?php print $title; ?></h3>
<?php endif; ?>
<div class="<?php print $class; ?>"<?php print $attributes; ?>>
  <?php foreach ($rows as $row_number => $columns): ?>
    <div <?php if ($row_classes[$row_number]) { print 'class="' . $row_classes[$row_number] .'"';  } ?>>
      <?php foreach ($columns as $column_number => $item): ?>
        <div <?php if ($column_classes[$row_number][$column_number]) { print 'class="' . $column_classes[$row_number][$column_number] .'"';  } ?>>
          <?php print $item; ?>
        </div>
      <?php endforeach; ?>
    </div>
  <?php endforeach; ?>
</div>
```

**Comment by wesruv** [comment]
Overall though I'm really digging the approach. The config form settings are great, and I love how it's coming together.

Let me know if/when you'd like me to pop in and help with the issues I said I could chip in on.

File: core/modules/views/templates/views.theme.inc (hunk 9 above)
Comment (docwilmot) [reply core/modules/views/templates/views.theme.inc:714]: I actually started with using core Bootstrap and also tried avoiding inline styles, but it seems unavoidable unless we limit the number of columns. FWIW D8 also uses inline styles.

We could apply a CSS style based on the number of columns, and define the width in CSS:
```
$rows[$row_id][$col_id]['col_style'] = 'col-number-' . $columns;
```
and CSS:
```
.col-number-5 { width: 20%;}
```
But we'd need a declaration for every possible column number, and would need to limit columns then.

If Bootstrap, we could add the Bootstrap classes to the colums easily. But for a 12-column grid like we have, once a user made 6 or more columns, we'd have to make all of them `col-md-1` because anything less wouldnt fit. i.e. 7 columns of `col-md-2` would be wider than one `col-md-12`.

Furthermore, if the user is insane and makes a grid with > 12 columns, Bootstrap doesn't account for this.

In short, inline styles or limit number of columns. Or JS.

Comment (jenlampton) [reply core/modules/views/templates/views.theme.inc:714]: Would it be reasonable to limit the number of columns to 1, 2, 3, 4, 6 or 12?

Comment (docwilmot) [reply core/modules/views/templates/views.theme.inc:714]: I think thats too limiting personally. I'd think a 5 or 7-column design isnt that rare?

Comment (docwilmot) [reply core/modules/views/templates/views.theme.inc:714]: Think photo gallery.
And another thing is this is replacing established functionality. People will expect to be able to define any number of grids if theyre transferring from Drupal.

**Comment by docwilmot** [comment]
> there isn't any padding in between the grid elements

I was wondering if we could make padding an option in the Views UI? Too much?

File: core/modules/views/templates/views.theme.inc (hunk 9 above)
Comment (jenlampton) [reply core/modules/views/templates/views.theme.inc:714]: > I think that's too limiting personally. I'd think a 5 or 7-column design isnt that rare?

I use 5 all the time (see my code above, I've added it to bootstrap) but I haven't ever seen 7.

> And another thing is this is replacing established functionality. People will expect to be able to define any number of grids if theyre transferring from Drupal.

True, but they are also likely going to be confused when they realize this is using a different grid system than the rest of their site.  Long term, I'd to love to move this onto bootstrap, but it doesn't need to be right now :)

**Comment by jenlampton** [comment]
>  Too much?

too much.

File: core/modules/views/templates/views.theme.inc (hunk 9 above)
Comment (docwilmot) [reply core/modules/views/templates/views.theme.inc:714]: I think adding additional classes for missing Bootstrap columns could do. Let's make a final decision then: 
- inline style or
- limit columns and
  - use Bootstrap with additional styles for 5, 7, 8, 9, 10 columns 
  - define customs styles

I'd like to hand this over to @wesruv to fix the CSS

**Comment by docwilmot** [comment]
I think this was closed accidentally. Awaiting a decision from @jenlampton @wesruv and anyone else who wants to chime in on https://github.com/backdrop/backdrop/pull/1784#discussion_r110288893

File: core/modules/views/templates/views.theme.inc (hunk 9 above)
Comment (docwilmot) [reply core/modules/views/templates/views.theme.inc:714]: @jenlampton @wesruv @quicksketch please advise. I'd like to get this wrapped up this weekend.

Comment (quicksketch) [reply core/modules/views/templates/views.theme.inc:714]: Inline styles seem reasonable here. It specifies the number of columns in the same way that tables would (but with the benefit of being responsive). Plus we avoid the dependency on loading Bootstrap (for themes/layouts that aren't using it).

File: core/modules/views/plugins/views_plugin_style_grid.inc [hunk 10]
```diff
@@ -31,41 +35,111 @@ class views_plugin_style_grid extends views_plugin_style {
    */
   function options_form(&$form, &$form_state) {
     parent::options_form($form, $form_state);
+    
+    // Flatten options to deal with the various hierarchy changes.
+    $options = views_grid_get_options($this->options);
```
Comment (quicksketch) [inline core/modules/views/plugins/views_plugin_style_grid.inc:40]: I suggest we make this a static method in this class rather than put it in views module. e.g.

```
static function flatten_options($options) {

}
```

Then call it with:

```
$options = views_plugin_style_grid::flatten_options($options);
```

Though honestly, perhaps it's better to just forgo flattening entirely and use the setting locations in the nested array.

File: core/modules/views/plugins/views_plugin_style_grid.inc [hunk 11]
```diff
@@ -31,41 +35,111 @@ class views_plugin_style_grid extends views_plugin_style {
...
       '#title' => t('Number of columns'),
-      '#default_value' => $this->options['columns'],
+      '#default_value' => $options['columns'],
       '#required' => TRUE,
-      '#min' => 1,
-      '#step' => 1,
+      '#element_validate' => array('views_element_validate_integer'),
+    );
+    $form['automatic_width'] = array(
+      '#type' => 'checkbox',
```
Comment (quicksketch) [inline core/modules/views/plugins/views_plugin_style_grid.inc:50]: There is one extra space on this indentation.

**Comment by klonos** [comment]
@docwilmot can you please close this? We have addressed this in https://github.com/backdrop/backdrop-issues/issues/2811
